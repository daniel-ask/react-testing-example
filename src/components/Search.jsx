import React, { useState } from "react";
import axios from "axios";
import SearchForm from "./SearchForm";

export default function Search() {
  const [data, setData] = useState([]);
  const [error, setError] = useState("");

  const retrieveGames = async (e, game) => {
    e.preventDefault();
    try {
      const response = await axios.get(
        `https://www.cheapshark.com/api/1.0/games?title=${game}&limit=60&exact=0`
      );
      if (response.data.length === 0) {
        setError("No Match Found");
      } else {
        setData(response.data);
      }
    } catch (error) {
      setError("Error");
    }
  };

  return (
    <div>
      <h1>Search for Game</h1>
      <SearchForm retrieveGames={retrieveGames} />
      {error && <h2>{error}</h2>}
      {data && (
        <ul>
          {data.map((game) => {
            return (
              <li key={game.gameID}>
                {game.external} cheapest: ${game.cheapest}
              </li>
            );
          })}
        </ul>
      )}
    </div>
  );
}
