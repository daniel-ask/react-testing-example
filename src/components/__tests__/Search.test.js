import { render, screen, cleanup, act } from "@testing-library/react";
import axios from "axios";
import userEvent from "@testing-library/user-event";
import Search from "../Search";
import renderer from "react-test-renderer";

jest.mock("axios");

describe("Search", () => {
  afterEach(cleanup);
  beforeEach(() => {render(<Search />)})

  test("should have the correct title", () => {
    expect(screen.getByText(/Search for Games/i)).toBeInTheDocument();
  });

  test("fetches games from an API and displays them", async () => {
    const games = [
      { gameID: 1, external: "batman", cheapest: 10.0 },
      { gameID: 2, external: "batman 2", cheapest: 5.0 },
    ];

    axios.get.mockImplementationOnce(() => Promise.resolve({ data: games }));

    await act(async () => {
      await userEvent.click(screen.getByRole("button"));
    });

    const items = await screen.findAllByRole("listitem");
    const message = screen.queryByText(/No Match Found/i);

    expect(items).toHaveLength(2);
    expect(message).toBeNull();
  });

  test("fetches games from the api but with no result", async () => {
    const games = [];
    axios.get.mockImplementationOnce(() => Promise.resolve({ data: games }));

    await act(async () => {
      await userEvent.click(screen.getByRole("button"));
    });
    const message = await screen.findByText(/No Match Found/i);
    expect(message).toBeInTheDocument();
  });

  test("fetches stories from an API and fails", async () => {
    axios.get.mockImplementationOnce(() => Promise.reject(new Error()));

    await act(async () => {
      await userEvent.click(screen.getByRole("button"));
    });

    const message = await screen.findByText(/Error/);

    expect(message).toBeInTheDocument();
  });

  it("should match last snapshot", () => {
    const tree = renderer.create(<Search />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});


