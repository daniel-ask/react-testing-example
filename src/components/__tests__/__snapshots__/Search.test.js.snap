// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should match last snapshot 1`] = `
<div>
  <h1>
    Search for Game
  </h1>
  <form
    onSubmit={[Function]}
  >
    <input
      id="game"
      name="game"
      onChange={[Function]}
      placeholder="Enter a game"
      type="text"
      value=""
    />
    <button
      type="submit"
    >
      Search for Games
    </button>
  </form>
  
  <ul />
</div>
`;
